README.md
# gh-issues-stats

[![NPM version][npm-image]][npm-url] [![Downloads][downloads-image]][npm-url] [![Build Status][travis-image]][travis-url] [![Coverage Status][codecov-image]][codecov-url] [![Dependency status][david-dm-image]][david-dm-url] [![Dev Dependency status][david-dm-dev-image]][david-dm-dev-url] [![Greenkeeper badge][greenkeeper-image]][greenkeeper-url]

[npm-url]:https://npmjs.org/package/gh-issues-stats
[downloads-image]:https://img.shields.io/npm/dm/gh-issues-stats.svg
[npm-image]:https://img.shields.io/npm/v/gh-issues-stats.svg
[travis-url]:https://travis-ci.org/moxystudio/node-gh-issues-stats
[travis-image]:http://img.shields.io/travis/moxystudio/node-gh-issues-stats/master.svg
[codecov-url]:https://codecov.io/gh/moxystudio/node-gh-issues-stats
[codecov-image]:https://img.shields.io/codecov/c/github/moxystudio/node-gh-issues-stats/master.svg
[david-dm-url]:https://david-dm.org/moxystudio/node-gh-issues-stats
[david-dm-image]:https://img.shields.io/david/moxystudio/node-gh-issues-stats.svg
[david-dm-dev-url]:https://david-dm.org/moxystudio/node-gh-issues-stats?type=dev
[david-dm-dev-image]:https://img.shields.io/david/dev/moxystudio/node-gh-issues-stats.svg
[greenkeeper-image]:https://badges.greenkeeper.io/moxystudio/node-gh-issues-stats.svg
[greenkeeper-url]:https://greenkeeper.io/

Collect statistical information about issues of a GitHub repository.


## Installation

```sh
$ npm install gh-issues-stats
```


## Usage

```js
const ghIssueStats = require('gh-issues-stats');

ghIssueStats('moxystudio/node-cross-spawn')
.then((stats) => {
    // `stats` looks like:
    // {
    //   issues: {
    //     count: 32,
    //     openCount: 2,
    //     distribution: {
    //       3600: 4,           // 1 hour
    //       10800: 20,         // 3 hours.. and so on
    //       32400: 6,
    //       97200: 2,
    //       291600: 0,
    //       874800: 0,
    //       2624400: 0,
    //       7873200: 0,
    //       23619600: 0,
    //       70858800: 0,
    //       212576400: 0,
    //     },
    //   },
    //   pullRequests: {
    //      .. same as above but for pull requests
    //   },
    // }
}, (err) => {
    console.log('Failed to fetch issue stats', err);
});
```

Available options:

- `apiUrl`: GitHub API URL, defaults to `https://api.github.com` (you may change to point to a GitHub enterprise instance).
- `tokens`: Array of API tokens to be used by [token-dealer](https://github.com/moxystudio/node-token-dealer), defaults to `null`.
- `concurrency`: The concurrency in which pages are requested, defaults to `5`.
- `got` Custom options to be passed to [got](https://github.com/sindresorhus/got), defaults to `{ timeout: 15000, headers: { accept: 'application/vnd.github.v3+json' } }`
- `tokenDealer`: Custom options to be passed to [token-dealer](https://github.com/moxystudio/node-token-dealer), defaults to `{ group: 'github' }`


## Tests

```sh
$ npm test
$ npm test -- --watch # during development
```


## License

Released under the [MIT License](http://www.opensource.org/licenses/mit-license.php).


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="3.0.1"></a>
## [3.0.1](https://github.com/moxystudio/node-gh-issues-stats/compare/v3.0.0...v3.0.1) (2018-09-23)



<a name="3.0.0"></a>
# 3.0.0 (2018-09-23)


### Chores

* upgrade project ([80089f0](https://github.com/moxystudio/node-gh-issues-stats/commit/80089f0))


### BREAKING CHANGES

* only node v8 or higher is supported



<a name="2.0.4"></a>
## 2.0.4 (2016-11-14)



<a name="2.0.3"></a>
## 2.0.3 (2016-09-18)



<a name="2.0.2"></a>
## 2.0.2 (2016-09-18)



<a name="2.0.1"></a>
## 2.0.1 (2016-09-14)



<a name="2.0.0"></a>
# 2.0.0 (2016-03-26)



<a name="1.0.0"></a>
# 1.0.0 (2016-03-20)



<a name="0.1.1"></a>
## 0.1.1 (2016-03-01)



<a name="0.1.0"></a>
# 0.1.0 (2016-02-29)


test/test.js
'use strict';

const https = require('https');
const nock = require('nock');
const nockBack = require('nock').back;
const expect = require('chai').expect;
const tokenDealer = require('token-dealer');
const ghIssueStats = require('../');

nockBack.fixtures = `${__dirname}/fixtures`;

const tokens = ['1e36f57b16d71d1df5c2188bb17c6c38f5ff206b'];  // Add your token here if you are recording fixtures, but do not commit

describe('gh-issues-stats', () => {
    const originalDateNow = Date.now;

    function mockDateNow() {
        Date.now = function () { return 1458477519537; };
    }

    afterEach(() => {
        Date.now = originalDateNow;

        nock.cleanAll();
        nock.disableNetConnect();
    });

    it('should give stats of a repository with a single issues page', () => {
        let nockBackDone;

        nockBack('gloth.json', (done) => { nockBackDone = done; });
        mockDateNow();

        return ghIssueStats('IndigoUnited/node-gloth', { tokens })
        .then((stats) => {
            nockBackDone();

            expect(stats).to.eql({
                issues: {
                    count: 0,
                    openCount: 0,
                    distribution: {
                        3600: 0,
                        10800: 0,
                        32400: 0,
                        97200: 0,
                        291600: 0,
                        874800: 0,
                        2624400: 0,
                        7873200: 0,
                        23619600: 0,
                        70858800: 0,
                        212576400: 0,
                    },
                },
                pullRequests: {
                    count: 8,
                    openCount: 0,
                    distribution: {
                        3600: 3,
                        10800: 1,
                        32400: 1,
                        97200: 0,
                        291600: 1,
                        874800: 2,
                        2624400: 0,
                        7873200: 0,
                        23619600: 0,
                        70858800: 0,
                        212576400: 0,
                    },
                },
            });
        });
    });

    it('should give stats of a repository with multiple issues page', () => {
        let nockBackDone;

        nockBack('nodejs.json', (done) => { nockBackDone = done; });
        mockDateNow();

        return ghIssueStats('nodejs/node', { tokens })
        .then((stats) => {
            nockBackDone();

            expect(stats).to.eql({
                issues: {
                    count: 3880,
                    openCount: 562,
                    distribution: {
                        3600: 610,
                        10800: 228,
                        32400: 253,
                        97200: 335,
                        291600: 337,
                        874800: 373,
                        2624400: 413,
                        7873200: 465,
                        23619600: 407,
                        70858800: 116,
                        212576400: 343,
                    },
                },
                pullRequests: {
                    count: 5116,
                    openCount: 260,
                    distribution: {
                        3600: 454,
                        10800: 250,
                        32400: 323,
                        97200: 666,
                        291600: 1129,
                        874800: 946,
                        2624400: 566,
                        7873200: 346,
                        23619600: 210,
                        70858800: 33,
                        212576400: 193,
                    },
                },
            });
        });
    });

    it('should use options.apiUrl', () => {
        nock('http://example.com')
          .get('/repos/my-org/my-repo/issues')
          .query(true)
          .reply(200, []);

        return ghIssueStats('my-org/my-repo', { apiUrl: 'http://example.com' })
        .then((stats) => {
            expect(stats.issues.count).to.equal(0);
            expect(stats.pullRequests.count).to.equal(0);
        });
    });

    it('should use options.tokens when fetching pages, preserving the accept header', () => {
        let headers;

        nock('https://api.github.com')
          .get('/repos/my-org/my-repo/issues')
          .query(true)
          .reply(200, function () {
              headers = this.req.headers;
              return [];
          });

        return ghIssueStats('my-org/my-repo', { tokens: ['my-token'] })
        .then(() => {
            expect(headers.authorization).to.equal('token my-token');
            expect(headers.accept).to.equal('application/vnd.github.v3+json');
        });
    });

    it('should respect options.concurrency', () => {
        let nockBackDone;

        nockBack('nodejs.json', (done) => { nockBackDone = done; });

        // Record concurrency
        let concurrency = 0;
        const concurrencyStack = [];
        const originalRequest = https.request;

        https.request = function (options, callback) {
            concurrency += 1;
            concurrencyStack.push(concurrency);

            return originalRequest.call(https, options, (res) => {
                res.on('end', () => { concurrency -= 1; });
                callback(res);
            });
        };

        return ghIssueStats('nodejs/node', { tokens, concurrency: 1 })
        .then(() => {
            nockBackDone();

            expect(concurrencyStack.length).to.be.greaterThan(10);
            concurrencyStack.forEach((concurrency) => expect(concurrency).to.equal(1));
        });
    });

    it('should pass options.got to got()', () => {
        let retriesCount = 0;
        const retries = () => { retriesCount += 1; return 0; };

        return ghIssueStats('my-org/my-repo', { got: { retries } })
        .then(() => {
            throw new Error('Should have failed');
        }, (err) => {
            expect(err.message).to.match(/not allow net/i);
            expect(retriesCount).to.equal(1);
        });
    });

    it('should pass options.tokenDealer to token-dealer', () => {
        nock('https://api.github.com')
          .get('/repos/my-org/my-repo/issues')
          .query(true)
          .reply(200, []);

        return ghIssueStats('my-org/my-repo', { tokens: ['my-token'], tokenDealer: { group: 'my-group' } })
        .then(() => {
            expect(tokenDealer.defaultLru.get('my-group#my-token')).to.be.an('object');
        });
    });

    it('should identify rate limit errors correctly', () => {
        const authorizationHeaders = [];

        nock('https://api.github.com')
          .get('/repos/my-org/my-repo/issues')
          .query(true)
          .reply(403, function () {
              authorizationHeaders.push(this.req.headers.authorization);

              return {
                  message: 'API rate limit exceeded',
                  documentation_url: 'https://developer.github.com/v3/#rate-limiting',
              };
          }, {
              'x-ratelimit-limit': '5000',
              'x-ratelimit-remaining': '0',
              'x-ratelimit-reset': Math.floor(Date.now() / 1000) + 10 * 60,
          });

        nock('https://api.github.com')
          .get('/repos/my-org/my-repo/issues')
          .query(true)
          .reply(200, function () {
              authorizationHeaders.push(this.req.headers.authorization);
              return [];
          }, {
              'x-ratelimit-limit': '5000',
              'x-ratelimit-remaining': '4999',
              'x-ratelimit-reset': Math.floor(Date.now() / 1000) + 60 * 60,
          });

        return ghIssueStats('my-org/my-repo', { got: { retries: 0 }, tokens: ['my-token', 'my-alternative-token'] })
        .then((stats) => {
            expect(stats.issues.count).to.equal(0);
            expect(stats.pullRequests.count).to.equal(0);
            expect(authorizationHeaders).to.eql(['token my-token', 'token my-alternative-token']);
        });
    });
});


test/index.spec.js
'use strict';

const https = require('https');
const nock = require('nock');
const nockBack = require('nock').back;
const tokenDealer = require('token-dealer');
const ghIssueStats = require('../');

nockBack.fixtures = `${__dirname}/fixtures`;

const tokens = ['1e36f57b16d71d1df5c2188bb17c6c38f5ff206b']; // Add your token here if you are recording fixtures, but do not commit

describe('gh-issues-stats', () => {
    const originalDateNow = Date.now;

    function mockDateNow() {
        Date.now = function () { return 1458477519537; };
    }

    beforeEach(() => {
        Date.now = originalDateNow;

        nock.cleanAll();
        nock.disableNetConnect();
    });

    it('should give stats of a repository with a single issues page', async () => {
        let nockBackDone;

        nockBack('gloth.json', (done) => { nockBackDone = done; });
        mockDateNow();

        const stats = await ghIssueStats('IndigoUnited/node-gloth', { tokens });

        nockBackDone();

        expect(stats).toEqual({
            issues: {
                count: 0,
                openCount: 0,
                distribution: {
                    3600: 0,
                    10800: 0,
                    32400: 0,
                    97200: 0,
                    291600: 0,
                    874800: 0,
                    2624400: 0,
                    7873200: 0,
                    23619600: 0,
                    70858800: 0,
                    212576400: 0,
                },
            },
            pullRequests: {
                count: 8,
                openCount: 0,
                distribution: {
                    3600: 3,
                    10800: 1,
                    32400: 1,
                    97200: 0,
                    291600: 1,
                    874800: 2,
                    2624400: 0,
                    7873200: 0,
                    23619600: 0,
                    70858800: 0,
                    212576400: 0,
                },
            },
        });
    });

    it('should give stats of a repository with multiple issues page', async () => {
        let nockBackDone;

        nockBack('nodejs.json', (done) => { nockBackDone = done; });
        mockDateNow();

        const status = await ghIssueStats('nodejs/node', { tokens });

        nockBackDone();

        expect(status).toEqual({
            issues: {
                count: 3880,
                openCount: 562,
                distribution: {
                    3600: 610,
                    10800: 228,
                    32400: 253,
                    97200: 335,
                    291600: 337,
                    874800: 373,
                    2624400: 413,
                    7873200: 465,
                    23619600: 407,
                    70858800: 116,
                    212576400: 343,
                },
            },
            pullRequests: {
                count: 5116,
                openCount: 260,
                distribution: {
                    3600: 454,
                    10800: 250,
                    32400: 323,
                    97200: 666,
                    291600: 1129,
                    874800: 946,
                    2624400: 566,
                    7873200: 346,
                    23619600: 210,
                    70858800: 33,
                    212576400: 193,
                },
            },
        });
    });

    it('should use options.apiUrl', async () => {
        nock('http://example.com')
        .get('/repos/my-org/my-repo/issues')
        .query(true)
        .reply(200, []);

        const stats = await ghIssueStats('my-org/my-repo', { apiUrl: 'http://example.com' });

        expect(stats.issues.count).toBe(0);
        expect(stats.pullRequests.count).toBe(0);
    });

    it('should use options.tokens when fetching pages, preserving the accept header', async () => {
        let headers = {};

        nock('https://api.github.com')
        .get('/repos/my-org/my-repo/issues')
        .query(true)
        /* eslint-disable no-invalid-this*/
        .reply(200, function () {
            headers = this.req.headers;

            return [];
        });
        /* eslint-enable no-invalid-this*/

        await ghIssueStats('my-org/my-repo', { tokens: ['my-token'] });

        expect(headers.authorization).toBe('token my-token');
        expect(headers.accept).toBe('application/vnd.github.v3+json');
    });

    it('should respect options.concurrency', async () => {
        let nockBackDone;

        nockBack('nodejs.json', (done) => { nockBackDone = done; });

        // Record concurrency
        let concurrency = 0;
        const concurrencyStack = [];
        const originalRequest = https.request;

        https.request = function (options, callback) {
            concurrency += 1;
            concurrencyStack.push(concurrency);

            return originalRequest.call(https, options, (res) => {
                res.on('end', () => { concurrency -= 1; });
                callback(res);
            });
        };

        await ghIssueStats('nodejs/node', { tokens, concurrency: 1 });

        nockBackDone();

        expect(concurrencyStack.length).toBeGreaterThan(10);
        concurrencyStack.forEach((concurrency) => expect(concurrency).toBe(1));
    });

    it('should pass options.got to got()', async () => {
        let retriesCount = 0;
        const retries = () => {
            retriesCount += 1;

            return 0;
        };

        expect.assertions(2);

        try {
            await ghIssueStats('my-org/my-repo', { got: { retry: { retries } } });
        } catch (err) {
            expect(err.message).toMatch(/disallowed net connect/i);
            expect(retriesCount).toBe(1);
        }
    });

    it('should pass options.tokenDealer to token-dealer', async () => {
        nock('https://api.github.com')
        .get('/repos/my-org/my-repo/issues')
        .query(true)
        .reply(200, []);

        await ghIssueStats('my-org/my-repo', { tokens: ['my-token'], tokenDealer: { group: 'my-group' } });

        expect(typeof tokenDealer.defaultLru.get('my-group#my-token')).toBe('object');
    });

    it('should identify rate limit errors correctly', async () => {
        const authorizationHeaders = [];

        nock('https://api.github.com')
        .get('/repos/my-org/my-repo/issues')
        .query(true)
        /* eslint-disable no-invalid-this, camelcase*/
        .reply(403, function () {
            authorizationHeaders.push(this.req.headers.authorization);

            return {
                message: 'API rate limit exceeded',
                documentation_url: 'https://developer.github.com/v3/#rate-limiting',
            };
        }, {
            'x-ratelimit-limit': '5000',
            'x-ratelimit-remaining': '0',
            'x-ratelimit-reset': Math.floor(Date.now() / 1000) + (10 * 60),
        });
        /* eslint-enable no-invalid-this, camelcase*/

        nock('https://api.github.com')
        .get('/repos/my-org/my-repo/issues')
        .query(true)
        /* eslint-disable no-invalid-this*/
        .reply(200, function () {
            authorizationHeaders.push(this.req.headers.authorization);

            return [];
        }, {
            'x-ratelimit-limit': '5000',
            'x-ratelimit-remaining': '4999',
            'x-ratelimit-reset': Math.floor(Date.now() / 1000) + (60 * 60),
        });
        /* eslint-enable no-invalid-this*/

        const stats = await ghIssueStats('my-org/my-repo', { got: { retries: 0 }, tokens: ['my-token', 'my-alternative-token'] });

        expect(stats.issues.count).toBe(0);
        expect(stats.pullRequests.count).toBe(0);
        expect(authorizationHeaders).toEqual(['token my-token', 'token my-alternative-token']);
    });
});


index.js
'use strict';

const url = require('url');
const got = require('got');
const pAll = require('p-all');
const tokenDealer = require('token-dealer');
const parseLink = require('github-parse-link');
const merge = require('lodash.merge');

const distributionRanges = [3600, 10800, 32400, 97200, 291600, 874800, 2624400, 7873200, 23619600, 70858800, 212576400];

function doRequest(url, options) {
    // Use token dealer to circumvent rate limit issues
    return tokenDealer(options.tokens, (token, exhaust) => {
        const handleRateLimit = (response, err) => {
            if (response.headers['x-ratelimit-remaining'] === '0') {
                const isRateLimitError = err && err.statusCode === 403 && /rate limit/i.test(response.body.message);

                exhaust(Number(response.headers['x-ratelimit-reset']) * 1000, isRateLimitError);
            }
        };

        return got(url, merge({}, options.got, {
            headers: token ? { authorization: `token ${token}` } : {},
            json: true,
        }))
        .then((response) => {
            handleRateLimit(response);

            return response;
        }, (err) => {
            err.response && handleRateLimit(err.response, err);
            throw err;
        });
    }, options.tokenDealer);
}

function getPagesAsArray(linkHeader) {
    const links = parseLink(linkHeader);
    const match = (links.last || '').match(/page=(\d+)$/);
    const totalPages = match ? Number(match[1]) : 0;
    const pages = [];

    for (let x = 1; x <= totalPages; x += 1) {
        pages.push(x);
    }

    return pages;
}

function parsePage(issues, stats) {
    issues.forEach((issue) => {
        const innerStats = issue.pull_request ? stats.pullRequests : stats.issues;

        // Update count
        innerStats.count += 1;

        // Update open count
        if (issue.state === 'open') {
            innerStats.openCount += 1;
        }

        // Update distribution count
        const closedTimestamp = (issue.closed_at ? Date.parse(issue.closed_at) : Date.now());
        const openTime = (closedTimestamp - Date.parse(issue.created_at)) / 1000;
        const rangeIndex = distributionRanges.findIndex((range, index, ranges) => {
            const previousRange = ranges[index - 1] || 0;

            return openTime >= previousRange && openTime < range;
        });
        /* istanbul ignore next */
        const range = distributionRanges[rangeIndex === -1 ? distributionRanges.length - 1 : rangeIndex];

        innerStats.distribution[range] += 1;
    });
}

function generateEmptyStats() {
    const stats = {
        issues: {
            count: 0,
            openCount: 0,
            distribution: {},
        },
        pullRequests: {
            count: 0,
            openCount: 0,
            distribution: {},
        },
    };

    distributionRanges.forEach((range) => {
        stats.issues.distribution[range] = 0;
        stats.pullRequests.distribution[range] = 0;
    });

    return stats;
}

// -------------------------------------------------

function ghIssueStats(repository, options) {
    options = merge({
        // GitHub API URL, you may change to point to a GitHub enterprise instance
        apiUrl: 'https://api.github.com',
        // Array of API tokens to be used by `token-dealer`
        tokens: null,
        // The concurrency in which pages are requested
        concurrency: 5,
        // Custom options to be passed to `got`
        got: {
            timeout: 15000,
            headers: { accept: 'application/vnd.github.v3+json' },
        },
        // Custom options to be passed to `token-dealer`
        tokenDealer: { group: 'github' },
    }, options);

    const stats = generateEmptyStats();
    const issuesUrl = url.resolve(options.apiUrl, `repos/${repository}/issues?state=all&per_page=100`);

    // Fetch first page
    return doRequest(issuesUrl, options)
    .then((response) => {
        parsePage(response.body, stats);

        // Fetch the remaining pages concurrently
        const remainingPages = getPagesAsArray(response.headers.link).slice(1);
        const actions = remainingPages.map((page) => () =>
            doRequest(`${issuesUrl}&page=${page}`, options)
            .then((response) => parsePage(response.body, stats))
        );

        return pAll(actions, { concurrency: options.concurrency });
    })
    .then(() => stats);
}

module.exports = ghIssueStats;
